pennylane/devices/default_clifford.py

```diff
@@ -0,0 +1,644 @@
+import concurrent.futures
+import numpy as np
+
+import pennylane as qml
+from pennylane.tape import QuantumTape, QuantumScript
+from pennylane.typing import Result, ResultBatch
+from pennylane.transforms import convert_to_numpy_parameters
+from pennylane.transforms.core import TransformProgram
+from pennylane.measurements import ExpectationMP, StateMP, DensityMatrixMP, PurityMP
+from pennylane.devices.qubit.sampling import get_num_shots_and_executions
+
+from . import Device
+from .execution_config import ExecutionConfig, DefaultExecutionConfig
+
+from .default_qubit import accepted_sample_measurement
+
+from .preprocess import (
+    decompose,
+    validate_observables,
+    validate_measurements,
+    validate_multiprocessing_workers,
+    validate_device_wires,
+    warn_about_trainable_observables,
+)
+
+Result_or_ResultBatch = Union[Result, ResultBatch]
+QuantumTape_or_Batch = Union[QuantumTape, Sequence[QuantumTape]]
+
+# Updated observable list
+_MEAS_OBSERVABLES = {
+    "PauliX",
+    "PauliY",
+    "PauliZ",
+    "Hadamard",
+    "Hermitian",
+    "Identity",
+    "Projector",
+    "Hamiltonian",
+    "Sum",
+    "SProd",
+    "Prod",
+    "Exp",
+}
+
+# Clifford gates
+_GATE_OPERATIONS = {
+    "Identity": "I",
+    "PauliX": "X",
+    "PauliY": "Y",
+    "PauliZ": "Z",
+    "Hadamard": "H",
+    "S": "S",
+    "Adjoint(S)": "S_DAG",
+    "SX": "SX",
+    "Adjoint(SX)": "SX_DAG",
+    "CNOT": "CNOT",
+    "SWAP": "SWAP",
+    "ISWAP": "ISWAP",
+    "Adjoint(ISWAP)": "ISWAP_DAG",
+    "CY": "CY",
+    "CZ": "CZ",
+    "GlobalPhase": None,
+    "BasisState": None,
+    "StatePrep": None,
+    "Snapshot": None,
+    "Barrier": None,
+}
+
+
+def operation_stopping_condition(op: qml.operation.Operator) -> bool:
+    """Specifies whether or not an operation is accepted by DefaultClifford."""
+    return op.name in _GATE_OPERATIONS
+
+
+def observable_stopping_condition(obs: qml.operation.Operator) -> bool:
+    """Specifies whether or not an observable is accepted by DefaultClifford."""
+    return obs.name in _MEAS_OBSERVABLES
+
+
+def _import_stim():
+    """Import stim."""
+    try:
+        # pylint: disable=import-outside-toplevel, unused-import, multiple-imports
+        import stim
+    except ImportError as Error:
+        raise ImportError(
+            "This feature requires stim, a fast stabilizer circuit simulator. "
+            "It can be installed with:\n\npip install stim"
+        ) from Error
+    return stim
+
+
+class DefaultClifford(Device):
+    r"""A PennyLane device written in Python and capable of executing Clifford circuits using
+    `stim (2021) <https://github.com/quantumlib/stim/tree/main>`_.
+
+    Args:
+        shots (int, Sequence[int], Sequence[Union[int, Sequence[int]]]): The default number of shots to use in executions involving
+            this device.
+        check_clifford (bool): Check if all the gate operations in the circuits to be executed are Clifford. Default is ``True``.
```

Comment:
I'm still a little on the fence about this argument. If `check_clifford=False` and you give a non-Clifford gate like `RX`, the error raised isn't actually that cryptic:
```
KeyError: 'RX'
```
When `check_clifford=True`, we get:
```
DeviceError: Operator RX(0.4, wires=[0]) not supported on default.clifford and does not provide a decomposition.
```

This is maybe clearer, but how much does this validation cost? Does it happen before execution or when that gate is encountered partway-through execution?

Reply:
I de-crypted the error message and now the same error message should appear in the case of `check_clifford=False` as well.

Reply:
I prefer to keep it for the reason we had added it in the PRD - allowing users to skip a redundant check iteration during the preprocessing and allowing us to keep the efficiency a bit closer to stim when possible.

Reply:
I don't mind keeping it, but not just because we already decided to - we can definitely revisit decisions together.

> allowing users to skip a redundant check iteration during the preprocessing and allowing us to keep the efficiency a bit closer to stim when possible

I guess my point is - can we just _always_ skip the checking that all gates are Clifford?

Reply:
As discussed, this is useful when we want to skip attempting to decompose circuit blocks into Clifford gates when the user promises to provide circuits with only Clifford gates.